buildSrc/subprojects/parser/src/main/kotlin/org/apache/calcite/buildtools/parser/DialectGenerateTask.kt

```diff
@@ -0,0 +1,210 @@
+     */
+    private fun getTraversalPath(rootDirectoryFile: File): Queue<String> {
+        val dialectDirectoryFile = dialectDirectory.get().asFile
+        var dialectPath = dialectDirectoryFile.absolutePath
+        val rootPath = rootDirectoryFile.absolutePath
+        val rootIndex = dialectPath.indexOf(rootPath)
+        dialectPath = dialectPath.substring(rootIndex + rootPath.length + 1)
+        return LinkedList(dialectPath.split("/"))
+    }
+
+    /**
+     * Traverses the determined path given by the queue. Once the queue is
+     * empty, the dialect directory has been reached. In that case any *.ftl
+     * file should be processed and no further traversal should happen.
+     *
+     * @param directories The directories to traverse in topdown order
+     * @param currentDirectory The current directory the function is processing
+     * @param functionMap The map to which the parsing functions will be added to
+     */
+    private fun traverse(
+        directories: Queue<String>,
+        currentDirectory: File,
+        functionMap: MutableMap<String, String>
+    ) {
+        val files = currentDirectory.listFiles()
+        // Ensures that files are processed first.
+        files.sortBy { it.isDirectory }
+        val nextDirectory = directories.peek()
+        for (f in files) {
+            if (f.isFile && f.extension == "ftl") {
+                processFile(f, functionMap)
+            }
+            if (directories.isNotEmpty() && f.name == nextDirectory) {
+                directories.poll()
+                traverse(directories, f, functionMap)
+            }
+        }
+    }
+
+    /**
+     * Extracts the functions from the given file into functionMap. Parses
+     * functions of the form:
+     * <return_type> <name>(<args>) :
+     * {
+     *     <properties>
+     * }
+     * {
+     *     <body>
+     * }
+     * @param file The file to process
+     * @param functionMap The map to which the parsing functions will be added to
+     */
+    private fun processFile(file: File, functionMap: MutableMap<String, String>) {
+        var fileText = file.readText(Charsets.UTF_8)
+        val typeAndName = "\\w+\\s+\\w+"
+        val declarationPattern =
+            Regex("(%s\\s*\\(\\s*(%s\\s*(\\,\\s*%s\\s*)*)?\\)\\s*\\:\n)"
+                    .format(typeAndName, typeAndName, typeAndName))
+        val delims = "(\\s|\n|\"|(//)|(/\\*)|(\\*/))"
+        // Uses Lookahead and Lookbehind to ensure that the delims are added as tokens.
+        val splitRegex = Regex("((?<=%s)|(?=%s))".format(delims, delims))
+        val declarationMatches = declarationPattern.findAll(fileText)
+        for (m in declarationMatches) {
+            val functionDeclaration = m.value
+            val functionName = getFunctionName(functionDeclaration)
+            val functionBuilder = StringBuilder(functionDeclaration)
+            val declarationIndex = fileText.indexOf(functionDeclaration)
```

Comment:
Performing a regex search and then following that up with another linear string search for every function seems expensive and could become problematic for large parsing files. In the worst case, I think this is O(N+N*M)=>O(N*M) where N is the number of characters in the file and M is the number of functions.

Are there any optimizations we can make here? For example, should we do away with the regex altogether and just iterate through the text ourselves? Then, we should be able to bring this down to just O(N).

Reply:
That should be possible. I think we can iterate over it ourselves but only start "parsing" when we found a function declaration. So we can keep a temp string variable and keep updating it until it matches the declaration regex.